### ISSUE TITLE
CCE returning AnyVal from concrete method calling super that abstracts over the AnyVal type

### ISSUE THREAD

**Issue by nafg** [scala/bug#12809]
## Reproduction steps

Scala version: Scala 2, including 2.13.11, but going way back too

Tested on JVM and JS. Confirmed it works on Scala 3 (both)

Code is at https://github.com/nafg/reproduce-anyval-cce

```scala
class Inner
class Outer(val inner: Inner) extends AnyVal
trait Abstract[F] {
  protected def helper: F
  protected def method(): F = helper
}
trait Concrete extends Abstract[Outer] {
  override protected def helper: Outer = new Outer(new Inner())
}
trait Public extends Concrete {
  def method2(): Outer = super.method()
}

object instance extends Public

instance.method2()
```

This is a minimized version of a real world problem. I discussed it at https://discord.com/channels/632150470000902164/635668814956068864/1119201503937302610

Translation:

| Name in minimized | Real class |
|--------|--------|
| `Inner` | `japgolly.scalajs.react.callback.Trampoline` |
| `Outer` | `japgolly.scalajs.react.callback.CallbackTo` |
| `Abstract` | `japgolly.scalajs.react.extra.BroadcasterF` |
| `Abstract#helper` | `japgolly.scalajs.react.util.Effect.UnsafeSync#traverse_` |
| `Abstract#method` | `japgolly.scalajs.react.extra.BroadcasterF#broadcast` |
| `Concrete` | `japgolly.scalajs.react.extra.Broadcaster` |
| `Concrete#helper` | `japgolly.scalajs.react.util.EffectCallback.callback#traverse_` |
| `Public` | `io.github.nafg.scalajs.react.util.PublicBroadcaster` | 

## Problem

It crashes at runtime with `Exception in thread "main" java.lang.ClassCastException: class Outer cannot be cast to class Inner`

Note that removing `super.` in `method2` fixes it. Originally I was overriding the same method just to make it public.

**Comment by nafg** [comment]
If there is a workaround for the case of `override def method(): Outer = super.method()` I'd appreciate knowing it.

**Comment by noresttherein** [comment]
I don't think it should compile at all, at least it would be consistent with how it works in general. In general, you can't provide a value class as a type parameter to a super class if that type is either returned from a method (directly, not wrapped), or used as a method argument. You get the annoying error of 'bridge method generated for <superclass.method> clashes with the method <subclass.method> itself because they both have the same signature after erasure'. Now, I don't think this is an unsolvable issue, but last time I inquired, nobody was in a hurry to change it.

**Comment by noresttherein** [comment]
If you'd like the same effect, you can go with 'poor man's opaque types` for Scala 2:
```Scala
class Inner
type Outer //left abstract

def Outer(inner :Inner) :Outer = inner.asInstanceOf[Outer]

implicit class OuterExtension(val self :Outer) extends AnyVal {
   def inner :Inner = self.asInstanceOf[Inner]
    //any methods which would go in a value class Outer go here.
}
```
This trick is very useful exactly because it avoids both your problem and the problem of bridge method clashing with specialized override in a subclass.

**Comment by nafg** [comment]
Are you suggesting to rewrite scalajs-react that way?
